XenAdmin/Wizards/PatchingWizard/PatchingWizard_AutoUpdatingPage.cs

```diff
@@ -96,7 +110,348 @@ public override bool EnableCancel()
+
+                progressBar.Value = 100;
+                _cancelEnabled = false;
+                _nextEnabled = true;
+            }
+            OnPageUpdated();
+
+            _thisPageHasBeenCompleted = true;
+        }
+
+        private void UpdateDelayedAfterPatchGuidanceActionListForHost(List<PlanAction> delayedGuidances, Host host, XenServerPatch patch)
+        {
+            List<PlanAction> actions = new List<PlanAction>();
+
+            List<XenRef<VM>> runningVMs = RunningVMs(host);
+
+            if (patch.after_apply_guidance == after_apply_guidance.restartHost)
+            {
+                actions.Add(new EvacuateHostPlanAction(host));
+                actions.Add(new RebootHostPlanAction(host));
+                actions.Add(new BringBabiesBackAction(runningVMs, host, false));
+            }
+
+            if (patch.after_apply_guidance == after_apply_guidance.restartXAPI)
+            {
+                actions.Add(new RestartAgentPlanAction(host));
+            }
+
+            if (patch.after_apply_guidance == after_apply_guidance.restartHVM)
+            {
+                actions.Add(new RebootVMsPlanAction(host, RunningHvmVMs(host)));
+            }
+
+            if (patch.after_apply_guidance == after_apply_guidance.restartPV)
+            {
+                actions.Add(new RebootVMsPlanAction(host, RunningPvVMs(host)));
+            }
+
+            if (!patch.GuidanceMandatory)
+            {
+                //not mandatory, so these actions will have to be run later
+                //add the ones that are not yet there
+                foreach (var ap in actions)
+                    if (!delayedGuidances.Any(da => da.GetType() == ap.GetType()))
+                    {
+                        //special rules
+                        //do not add restartXAPI if there is already a restartHost on the list
+                        if (delayedGuidances.Any(da => da is RebootHostPlanAction) && ap is RestartAgentPlanAction)
+                            continue;
+
+                        delayedGuidances.Add(ap);
+                    }
+            }
+            else
+            {
+                //remove delayed action of the same kind for this host (because it is mandatory and will run immediately)
+                foreach (var mg in delayedGuidances)
+                    delayedGuidances.RemoveAll(a => a.GetType() == mg.GetType());
+
+                //if it is a restart, clean delayed list
+                if (patch.after_apply_guidance == after_apply_guidance.restartHost)
+                    delayedGuidances.Clear();
+                
+            }
+        }
+
+        private List<PlanAction> GetMandatoryActionListForPatch(List<PlanAction> delayedGuidances, Host host, XenServerPatch patch)
+        {
+            var actions = new List<PlanAction>();
+
+            if (!patch.GuidanceMandatory)
+                return actions;
+
+            List<XenRef<VM>> runningVMs = RunningVMs(host);

```

Comment:
This code (up to the end of the function) can be extracted into a new function and reused in the UpdateDelayedAfterPatchGuidanceActionListForHost as well